From b9d0da5328a4269315d2e93ec3f6083411c1b09b Mon Sep 17 00:00:00 2001
Message-Id: <b9d0da5328a4269315d2e93ec3f6083411c1b09b.1421830309.git.chang-joon.lee@intel.com>
In-Reply-To: <49dde4951ccd12ce81e60aed3b24793e0ac12fed.1421830308.git.chang-joon.lee@intel.com>
References: <49dde4951ccd12ce81e60aed3b24793e0ac12fed.1421830308.git.chang-joon.lee@intel.com>
From: Yogesh Mohan Marimuthu <yogesh.mohan.marimuthu@intel.com>
Date: Wed, 22 Oct 2014 16:46:20 +0530
Subject: [PATCH 15/17] FOR_UPSTREAM [VPG]: drm/i915: Add error check in
 skipping i2c sequence

The i2c sequence logic skipping doesn't reduce the tmp variable which
is used for error checking. fix it in this patch.

Issue: GMINL-1644
Change-Id: I3810161879542c7306dfaac56c96ad9064045e89
Signed-off-by: Yogesh Mohan Marimuthu <yogesh.mohan.marimuthu@intel.com>
---
 drivers/gpu/drm/i915/intel_bios.c |    8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/gpu/drm/i915/intel_bios.c b/drivers/gpu/drm/i915/intel_bios.c
index 924de46..0025a04 100644
--- a/drivers/gpu/drm/i915/intel_bios.c
+++ b/drivers/gpu/drm/i915/intel_bios.c
@@ -727,8 +727,16 @@ static u8 *goto_next_sequence(u8 *data, int *size)
 			break;
 		case MIPI_SEQ_ELEM_I2C:
 			/* skip by this element payload size */
+			tmp -= 8;
+			if (tmp < 0)
+				return NULL;
 			data += 7;
 			len = *data;
+
+			tmp -= len;
+			if (tmp < 0)
+				return NULL;
+
 			data += len + 1;
 			break;
 		default:
-- 
1.7.9.5

